controllers/kubeadmconfig_controller.go

```diff
@@ -263,6 +267,83 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 	return ctrl.Result{}, nil
 }
 
+// reconcileDiscovery ensure that config.JoinConfiguration.Discovery is properly set for the joining node.
+// The implementation func respect user provided discovery configurations, but in case some of them are missing, a valid BootstrapToken object
+// is automatically injected into config.JoinConfiguration.Discovery.
+// This allows to simplify configuration UX, by providing the option to delegate to CABPK the configuration of kubeadm join discovery.
+func (r *KubeadmConfigReconciler) reconcileDiscovery(cluster *capiv1alpha2.Cluster, config *cabpkv1alpha2.KubeadmConfig) error {
+	log := r.Log.WithValues("kubeadmconfig", fmt.Sprintf("%s/%s", config.Namespace, config.Name))
+
+	// if config already contains a file discovery configuration, respect it without further validations
+	if config.Spec.JoinConfiguration.Discovery.File != nil {
+		return nil
+	}
+
+	// otherwise it is necessary to ensure token discovery is properly configured
+	if config.Spec.JoinConfiguration.Discovery.BootstrapToken == nil {
+		config.Spec.JoinConfiguration.Discovery.BootstrapToken = &kubeadmv1beta1.BootstrapTokenDiscovery{}
+	}
+
+	// if BootstrapToken already contains an APIServerEndpoint, respect it; otherwise inject the APIServerEndpoint endpoint defined in cluster status
+	//TODO(fp) might be we want to validate user provided APIServerEndpoint and warn/error if it doesn't match the api endpoint defined at cluster level
+	apiServerEndpoint := config.Spec.JoinConfiguration.Discovery.BootstrapToken.APIServerEndpoint
+	if apiServerEndpoint == "" {
+		if len(cluster.Status.APIEndpoints) == 0 {
+			return errors.New("Control plane already exists for the cluster but cluster.Status.APIEndpoints are not set")
+		}
+
+		// NB. CABPK only uses the first APIServerEndpoint defined in cluster status if there are multiple defined.
+		apiServerEndpoint = fmt.Sprintf("https://%s:%d", cluster.Status.APIEndpoints[0].Host, cluster.Status.APIEndpoints[0].Port)
+		config.Spec.JoinConfiguration.Discovery.BootstrapToken.APIServerEndpoint = apiServerEndpoint
+		log.Info("Altering JoinConfiguration.Discovery.BootstrapToken", "APIServerEndpoint", apiServerEndpoint)
+	}
+
+	// if BootstrapToken already contains a token, respect it; otherwise create a new bootstrap token for the node to join
+	if config.Spec.JoinConfiguration.Discovery.BootstrapToken.Token == "" {
+		// if there is no secretInterfaceFactory in place
+		if r.secretInterfaceFactory == nil {
+			// create the default secretInterfaceFactory function, that connects to the given remote cluster
+			// and then gets the secretInterface client for the kube-system namespace
+			r.secretInterfaceFactory = func(cluster *capiv1alpha2.Cluster) (corev1.SecretInterface, error) {
```

Comment:
This function is only closing around `r`. It takes cluster as a parameter and everything else is calculated. If you define this function on the `KubeadmConfigReconciler` it no longer has to be a function and this nil and anonymous function can go away.

If you're interested in unit testing this behavior I would recommend something like this

```go
type secretsClientFactory interface {
    secretsClient(*client.Client, *v1alpha2.Cluster)
}

type KubeadmConfigReconciler struct {
    ...
    secretsClientFactory secretsClientFactory
}
```

I'd be happy to discuss this on zoom if you'd like.

Reply:
Discussed offline. We talked about where this should be initialized, I suggested the same point as the client for the reconciler as that's the only real dependency of this function.

Reply:
Thanks for helping me in shaping this out!